data/transactions/logic/assembler.go

```diff
@@ -0,0 +1,878 @@
+	"strconv"
+	"strings"
+
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+// Writer is what we want here. Satisfied by bufio.Buffer
+type Writer interface {
+	Write([]byte) (int, error)
+	WriteByte(c byte) error
+}
+
+type labelReference struct {
+	sourceLine int
+
+	// position of the opcode start that refers to the label
+	position int
+
+	label string
+}
+
+// OpStream is destination for program and scratch space
+type OpStream struct {
+	Out     bytes.Buffer
+	vubytes [9]byte
+	intc    []uint64
+	bytec   [][]byte
+
+	// Keep a stack of the types of what we would push and pop to typecheck a program
+	typeStack []StackType
+
+	// current sourceLine during assembly
+	sourceLine int
+
+	// map label string to position within Out buffer
+	labels map[string]int
+
+	labelReferences []labelReference
+}
+
+// SetLabelHere inserts a label reference to point to the next instruction
+func (ops *OpStream) SetLabelHere(label string) {
+	if ops.labels == nil {
+		ops.labels = make(map[string]int)
+	}
+	ops.labels[label] = ops.Out.Len()
+}
+
+// ReferToLabel records an opcode label refence to resolve later
+func (ops *OpStream) ReferToLabel(sourceLine, pc int, label string) {
+	ops.labelReferences = append(ops.labelReferences, labelReference{sourceLine, pc, label})
+}
+
+func (ops *OpStream) tpush(argType StackType) {
+	ops.typeStack = append(ops.typeStack, argType)
+}
+
+func (ops *OpStream) tpop() (argType StackType) {
+	if len(ops.typeStack) == 0 {
+		argType = StackNone
+		return
+	}
+	last := len(ops.typeStack) - 1
+	argType = ops.typeStack[last]
+	ops.typeStack = ops.typeStack[:last]
+	return
+}
+
+// Intc writes opcodes for loading a uint64 constant onto the stack.
+func (ops *OpStream) Intc(constIndex uint) error {
+	switch constIndex {
+	case 0:
+		ops.Out.WriteByte(0x22) // intc_0
+	case 1:
+		ops.Out.WriteByte(0x23) // intc_1
+	case 2:
+		ops.Out.WriteByte(0x24) // intc_2
+	case 3:
+		ops.Out.WriteByte(0x25) // intc_3
+	default:
+		if constIndex > 0xff {
+			return errors.New("cannot have more than 256 int constants")
+		}
+		ops.Out.WriteByte(0x21) // intc
+		ops.Out.WriteByte(uint8(constIndex))
+	}
+	ops.tpush(StackUint64)
+	return nil
+}
+
+// Uint writes opcodes for loading a uint literal
+func (ops *OpStream) Uint(val uint64) error {
+	found := false
+	var constIndex uint
+	for i, cv := range ops.intc {
+		if cv == val {
+			constIndex = uint(i)
+			found = true
+			break
+		}
+	}
+	if !found {
+		constIndex = uint(len(ops.intc))
+		ops.intc = append(ops.intc, val)
+	}
+	return ops.Intc(constIndex)
+}
+
+// Bytec writes opcodes for loading a []byte constant onto the stack.
+func (ops *OpStream) Bytec(constIndex uint) error {
+	switch constIndex {
+	case 0:
+		ops.Out.WriteByte(0x28) // bytec_0
+	case 1:
+		ops.Out.WriteByte(0x29) // bytec_1
+	case 2:
+		ops.Out.WriteByte(0x2a) // bytec_2
+	case 3:
+		ops.Out.WriteByte(0x2b) // bytec_3
+	default:
+		if constIndex > 0xff {
+			return errors.New("cannot have more than 256 byte constants")
```

Comment:
`"cannot have more than 256 byte constants")` - but the user can work around that limitation using byte arrays, right ?